colossalai/nn/optimizer/distributed_came.py

```diff
@@ -0,0 +1,307 @@
+        # record working parameter original shape (Before TP)
+        for group in self.param_groups:
+            for p in group["params"]:
+                if hasattr(p, "data"):
+                    self.ori_shape[id(p)] = p.data.shape
+                elif hasattr(p, "shape"):
+                    self.ori_shape[id(p)] = p.shape
+                else:
+                    self.ori_shape[id(p)] = p.size()
+
+    @property
+    def supports_memory_efficient_fp16(self):
+        return True
+
+    @property
+    def supports_flat_params(self):
+        return False
+
+    def setup_distributed(
+        self, tp_group: ProcessGroup, zero_group: ProcessGroup, master_to_working_map: dict, zero_flag: bool = False
+    ):
+        self.tensor_parallel_group = tp_group
+        self.zero_parallel_group = zero_group
+        # When running setup_distribute(), the parameters now are master parameters (After TP and zero)
+        # But master param != d_tensor, need to record information of working param (After TP before zero)
+        for group in self.param_groups:
+            for p in group["params"]:
+                # if no zero, working param == master param
+                working_param = master_to_working_map[id(p)] if master_to_working_map else p
+                self.ori_shape[id(p)] = self.ori_shape[id(working_param)]
+                self.working_shape[id(p)] = working_param.size()
+                self.gather_before_compute[id(p)] = False
+                try:
+                    # api.get_device_mesh(working_param)
+                    sharding_spec = api.get_sharding_spec(working_param)
+                    self.clip_method[id(p)] = "col" if 0 in sharding_spec.dim_partition_dict.keys() else "row"
+                    # if dist.get_rank() == 0:
+                    #     print(self.ori_shape[id(p)], p.data.shape, sharding_spec.dim_partition_dict, self.clip_method[id(p)])
+                except:
+                    self.clip_method[id(p)] = None
+
+        self.zero = True if zero_flag else False
+        self.distributed = True
+        # print(self.distributed, self.zero, dist.get_world_size(zero_group))
+
+    def _get_options(self, param_shape):
+        factored = len(param_shape) >= 2
+        return factored
+
+    def _rms(self, tensor, param):
+        if not self.distributed:
+            return tensor.norm(2) / (tensor.numel() ** 0.5)
+        # return tensor.norm(2) / (tensor.numel() ** 0.5)
+        # Calculate the sum of the squares of the tensors on the current device
+        sum_sq = tensor.pow(2).sum()
+
+        # Summing up the sum of squares across all devices
+        dist.all_reduce(sum_sq, op=dist.ReduceOp.SUM, group=self.tensor_parallel_group)
+
+        # Summarize the total number of elements across all devices
+        numel = torch.tensor(tensor.numel(), device=tensor.device)
+        dist.all_reduce(numel, op=dist.ReduceOp.SUM, group=self.tensor_parallel_group)
```

Comment:
I think we can get numel from working param, right? Communication is too expensive